File: algorithms/medium/3_longest_substring_without_repeating_characters.py
# 3. Longest Substring Without Repeating Characters

def length_of_longest_substring(s):
    """Find the length of the longest substring without repeating characters.
    
    Args:
        s: String to process
    
    Returns:
        Length of the longest substring without repeating characters
    """
    # Edge case: empty string
    if not s:
        return 0
    
    # Dictionary to store the last position of each character
    char_position = {}
    
    # Initialize variables to track the current substring
    start = 0
    max_length = 0
    
    # Iterate through the string
    for i, char in enumerate(s):
        # If the character is already in the current substring,
        # update the start position to exclude the previous occurrence
        if char in char_position and start <= char_position[char]:
            start = char_position[char] + 1
        else:
            # Update max_length if the current substring is longer
            max_length = max(max_length, i - start + 1)
        
        # Update the last position of the character
        char_position[char] = i
    
    return max_length

# Test cases
print(length_of_longest_substring("abcabcbb"))  # 3 ("abc")
print(length_of_longest_substring("bbbbb"))     # 1 ("b")
print(length_of_longest_substring("pwwkew"))    # 3 ("wke")
print(length_of_longest_substring(""))          # 0 (empty string)

File: algorithms/medium/3_longest_substring_without_repeating_characters.js
/**
 * 3. Longest Substring Without Repeating Characters
 * 
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    // Edge case: empty string
    if (!s) return 0;
    
    // Map to store the last position of each character
    const charPosition = new Map();
    
    // Initialize variables to track the current substring
    let start = 0;
    let maxLength = 0;
    
    // Iterate through the string
    for (let i = 0; i < s.length; i++) {
        const char = s[i];
        
        // If the character is already in the current substring,
        // update the start position to exclude the previous occurrence
        if (charPosition.has(char) && start <= charPosition.get(char)) {
            start = charPosition.get(char) + 1;
        } else {
            // Update maxLength if the current substring is longer
            maxLength = Math.max(maxLength, i - start + 1);
        }
        
        // Update the last position of the character
        charPosition.set(char, i);
    }
    
    return maxLength;
};

// Test cases
console.log(lengthOfLongestSubstring("abcabcbb"));  // 3 ("abc")
console.log(lengthOfLongestSubstring("bbbbb"));     // 1 ("b")
console.log(lengthOfLongestSubstring("pwwkew"));    // 3 ("wke")
console.log(lengthOfLongestSubstring(""));          // 0 (empty string)